index.js

```diff
@@ -61,15 +63,24 @@ module.exports = declare((api, options) => {
         loose: true,
       }] : null,
 
+      looseComputedProperties ? [require('@babel/plugin-transform-computed-properties'), {
+        loose: true,
+      }] : null,
+
+      looseParameters ? [require('@babel/plugin-transform-parameters'), {
+        loose: true,
+      }] : null,
+
+      looseTemplateLiterals ? [require('@babel/plugin-transform-template-literals'), {
+        loose: true,
```

Comment:
We shouldn't use this option or even permit it; Set vs Define is nbd in the other ones, but with loose mode, this doesn't properly stringify (because it uses `+` which prefers valueOf)

Reply:
@ljharb Can you explain this a little more to me? Specifically this part:
> this doesn't properly stringify (because it uses + which prefers valueOf)

Reply:
ah, I think I figured it out with this code:

```js
function Foo() {}
Foo.prototype.valueOf = function() {
  return 'ohai';
};

const a = new Foo();
console.log('' + a); // ohai
console.log(''.concat(a)); // [object Object]
```

Reply:
The snippet I usually use to demonstrate is this:
```js
var x = { toString: function () { return NaN; }, valueOf: function () { return Infinity; }}; [x.toString(), x + '', String(x)]
//=> [NaN, 'Infinity', 'NaN']
/* note that only the third item is both a string, and the toString value */
```

Reply:
Here's a relevant issue I found: https://github.com/babel/babel/issues/9743

Do you think it would make sense to propose changing the loose mode to compile like this? (Or perhaps adding a new mode that is somewhere in between loose and spec?)

```js
const foo = '' + String(a) + String(b);
```

It looks like this is how the spec mode used to work until it was changed in https://github.com/babel/babel/pull/5791 to support Symbol better.

For more context here, I'm looking to reduce the memory cost of how we compile template literals. Trying out the current spec compilation (`.concat()`) vs the proposed loose compilation (`'' + String(foo)`) still gives us a nice memory usage reduction

FWIW [there is an old microbenchmark](https://github.com/babel/babel/issues/1065) that indicates it may not be as fast as the current loose mode compilation, but not sure if that holds up in real world scenarios. Also, I [added a couple of other cases that show it `String(foo)` isn't that far off from `'' + foo`](https://jsperf.com/tostring-vs-appending-empty-string/5). All caveats of the usefulness of microbenchmarks apply, of course.

More concretely, here's a part of function that I'm doing some memory profiling on. The function is named `buildCSSVariable` and it returns a string using template literal syntax. The template literal compiled in strict mode:

```js
return `var(--${dashify(keys.join('-'))}, ${defaultValue})`;
```

Profiling using Chrome's allocation sampling, I recorded this profile (note that this function is called as part of a recursive operation so you'll see it show up in a few places in the stack):

![image](https://user-images.githubusercontent.com/195534/56473236-090a2600-641d-11e9-9274-46b055fd6bf5.png)

And here's the version I wrote by hand to approximate how we might compile this:

```js
return 'var(--' + String(dashify(keys.join('-'))) + ', ' + String(defaultValue) + ')';
```

![image](https://user-images.githubusercontent.com/195534/56473248-2c34d580-641d-11e9-8426-efbd4ee9f36d.png)

For completeness, here's a memory profile with the current loose mode compilation I wrote in by hand. Code:

```js
return 'var(--' + dashify(keys.join('-')) + ', ' + defaultValue + ')';
```

![image](https://user-images.githubusercontent.com/195534/56473386-a403ff80-641f-11e9-9ac6-598c5f4f9c14.png)

Reply:
I'd be open to an option that enables any alternative transformation to babel's spec mode that remains spec compliant - but if we can come up with that, then we should just PR that into babel's spec mode ourselves :-)

Reply:
Yeah, good point. I think joining an array might be a winner actually.

Here's the code I profiled:

```js
return ['var(--', dashify(keys.join('-')), ', ', defaultValue, ')'].join('');
```

And here's the same memory profile as above:
![image](https://user-images.githubusercontent.com/195534/56481321-531df680-6473-11e9-844d-f27953bbc478.png)

[The microbenchmark, FWIW, shows it as faster than `''.concat(foo)` too](https://jsperf.com/tostring-vs-appending-empty-string/7):

![image](https://user-images.githubusercontent.com/195534/56481378-8e202a00-6473-11e9-800c-47ad745a325e.png)

And, I think it complies with the spec as far as I know (although I think you're a far better judge of that than I am). Here are some test cases:

```js
var x = { toString: function () { return NaN; }, valueOf: function () { return Infinity; }};
[x].join(''); // "NaN"

var a = Symbol();
[a].join(''); // Uncaught TypeError: Cannot convert a Symbol value to a string
```

Are there other situations I should try out or should I bring this up on the Babel side?

Reply:
Relevant: https://docs.google.com/document/d/1o-MJPAddpfBfDZCkIHNKbMiM86iDFld7idGbNQLuKIQ/preview

Reply:
Looks good to me; improving the perf of spec options is a win for everyone!

Reply:
https://twitter.com/bmeurer/status/1120179461851361280